File: src/engraft/actions/regex_replace.py
import re
from dataclasses import dataclass, field
from pathlib import Path

from engraft.actions import register
from engraft.actions.base import Action


@register("regex_replace")
@dataclass
class RegexReplace(Action):
    """Replace values in text files using regex selectors with named capture groups."""

    file: str
    replace: list[dict[str, str]] = field(default_factory=list)

    def apply(
        self,
        variables: dict[str, str],
        work_dir: Path,
        values_dir: Path,
    ) -> None:
        target = work_dir / self.file
        content = target.read_text()

        for entry in self.replace:
            pattern = entry["selector"]
            var_name = entry["variable"]

            if "(?P<value>" not in pattern:
                raise ValueError(
                    f"Pattern must contain a (?P<value>...) named group: {pattern!r}"
                )

            new_value = variables[var_name]
            compiled = re.compile(pattern)

            if not compiled.search(content):
                raise ValueError(
                    f"Pattern {pattern!r} did not match anything in {self.file}"
                )

            def replacer(match: re.Match, new_val: str = new_value) -> str:
                start, end = match.span("value")
                full_start = match.start()
                prefix = match.group()[: start - full_start]
                suffix = match.group()[end - full_start :]
                return prefix + new_val + suffix

            content = compiled.sub(replacer, content)

        target.write_text(content)

    def target_files(self) -> list[str]:
        """Return project-relative file paths this action operates on."""
        return [self.file]

File: src/actions/regexReplace.ts
import { promises as fs } from "node:fs";
import path from "node:path";

import type { Action, Variables } from "./base.js";
import { register } from "./registry.js";

interface ReplaceEntry {
  selector: string;
  variable: string;
}

@register("regex_replace")
export class RegexReplace implements Action {
  private readonly file: string;
  private readonly replace: ReplaceEntry[];

  constructor(config: Record<string, unknown>) {
    this.file = String(config.file);
    this.replace = (config.replace as ReplaceEntry[] | undefined) ?? [];
  }

  async apply(variables: Variables, workDir: string, _valuesDir: string): Promise<void> {
    const target = path.join(workDir, this.file);
    let content = await fs.readFile(target, "utf8");

    for (const entry of this.replace) {
      const newValue = variables[entry.variable];
      if (newValue == null) continue;

      const normalized = entry.selector.replace(/\(\?P<value>/g, "(?<value>");

      if (!normalized.includes("(?<value>")) {
        throw new Error(
          `Pattern must contain a (?P<value>...) or (?<value>...) named group: ${JSON.stringify(entry.selector)}`,
        );
      }

      // The `d` flag records where each group matched. Searching the match
      // text for the captured string instead would rewrite the wrong run
      // whenever the same text also appears in the pattern's prefix, as in
      // `version="(?<value>[^"]*)"` against `version="version"`.
      const pattern = new RegExp(normalized, "dg");
      const matches = [...content.matchAll(pattern)];

      if (matches.length === 0) {
        throw new Error(
          `Pattern ${JSON.stringify(entry.selector)} did not match anything in ${this.file}`,
        );
      }

      // Splice from the end so each replacement leaves the offsets of the
      // ones still to come untouched.
      for (const match of matches.toReversed()) {
        const span = match.indices?.groups?.["value"];
        if (!span) continue;
        const [start, end] = span;
        content = content.slice(0, start) + newValue + content.slice(end);
      }
    }

    await fs.writeFile(target, content);
  }

  targetFiles(): string[] {
    return [this.file];
  }
}
